kanalony-web-server/app/controllers/Application.scala

```diff
@@ -0,0 +1,126 @@
+        println(e.getClass.getTypeName)
+        return wrapWithPromise(InternalServerError(e.getMessage))
+      }
+    }
+  }
+
+  def wrapWithPromise(res : Result): Future[Result] = {
+    return (Promise[Result] success res).future
+  }
+
+  def execute(queryParams: QueryParams): Future[IQueryResult] = {
+    QueryLocator.locate(queryParams).query(queryParams)
+  }
+
+  def extractValues[T, E <: Exception](values : List[String], enumConverter : String => T, exceptionCreator : String => E): List[T] = {
+    var processedValue = ""
+    var res = List[T]()
+    try {
+      values.foreach(d => {
+        processedValue = d
+        res = res :+ enumConverter(d)
+      })
+      res
+    }
+    catch {
+      case e : Exception => throw exceptionCreator(processedValue)
+    }
+  }
+
+  def extractMetrics(metrics : List[String]): List[Metrics.Value] = {
+    extractValues(metrics, Metrics.withName(_), s => new InvalidMetricsException("Metric " + s + " not supported"))
+  }
+
+  def extractDimension(dimension : String): Dimensions.Value = {
+    val results = extractValues(List(dimension), Dimensions.withName(_), s => new InvalidDimensionException(s))
+    results(0)
+  }
+
+  def extractDimensions(dimensions : List[String]): List[Dimensions.Value] = {
+    extractValues(dimensions, Dimensions.withName(_), s => new InvalidDimensionException(s))
+  }
+
+  def createConstraint(dimension: Dimensions.Value, values: List[String]) : IDimensionConstraint = {
+    try {
+      dimension match {
+        case Dimensions.partner => new DimensionEqualityConstraint[Int](values.toSet.map { (v: String) => v.toInt })
+        case _ => new DimensionEqualityConstraint(values.toSet)
+      }
+    }
+    catch {
+      case e : NumberFormatException => { throw new IllegalArgumentException("Dimension " + dimension + " supplied invalid values")}
+    }
+  }
+
+  def requestToQueryParams(req : AnalyticsRequest) : QueryParams = {
+
+    val dimensionsInResult = extractDimensions(req.dimensions)
+    val metricsInResult = extractMetrics(req.metrics)
+
+    if (metricsInResult.isEmpty){
+      throw new MetricNotSuppliedException
+    }
+
+    val constrainedDimensionDefinitions = req.filters.map {
+      f => {
+        val dimension = extractDimension(f.dimension)
+        val shouldAppearInResult = dimensionsInResult contains dimension
+        val dimensionConstraint = createConstraint(dimension, f.values)
+        QueryDimensionDefinition(dimension, dimensionConstraint, shouldAppearInResult)
+      }
+    }
+
+    val unconstrainedDimensionDefinitions = dimensionsInResult.filter(d => !(constrainedDimensionDefinitions.map(_.dimension) contains d)) map {
+      dimension => QueryDimensionDefinition(dimension , new DimensionUnconstrained, true)
+    }
+
+    // TODO : currently only a single metric is supported
+    QueryParams(constrainedDimensionDefinitions ::: unconstrainedDimensionDefinitions, metricsInResult(0), new DateTime(req.from), new DateTime(req.to))
```

Comment:
Dates are already validated as part of the AnalyticsRequest object creation. I need to see how we can understand what went wrong if there's a parsing error